core/src/main/java/com/linecorp/armeria/client/ClientHttpObjectEncoder.java

```diff
@@ -32,17 +33,20 @@ interface ClientHttpObjectEncoder extends HttpObjectEncoder {
     /**
      * Writes a {@link RequestHeaders}.
      */
-    default ChannelFuture writeHeaders(int id, int streamId, RequestHeaders headers, boolean endStream) {
+    default ChannelFuture writeHeaders(int id, int streamId, RequestHeaders headers, boolean endStream,
+                                       ChannelPromise promise) {
         assert eventLoop().inEventLoop();
         if (isClosed()) {
-            return newFailedFuture(UnprocessedRequestException.of(ClosedSessionException.get()));
+            promise.setFailure(UnprocessedRequestException.of(ClosedSessionException.get()));
```

Comment:
Maybe better using `tryFailure()` so we are totally sure we don't raise an exception here?
```suggestion
            promise.tryFailure(UnprocessedRequestException.of(ClosedSessionException.get()));
```

Reply:
Good suggestion!